### PR TITLE
chore: refactor BlobReadChannel to plumb the sparse object metadata included in a read media response for both json and grpc

### PR DESCRIPTION
**Pull request by BenWhitehead**
Publicly exposing this is gated on how/when we are able to expose StorageReadChannel.

### CODE REVIEW

File: google-cloud-storage/src/main/java/com/google/cloud/storage/GapicUnbufferedReadableByteChannel.java [hunk 1]
```diff
@@ -237,12 +237,26 @@ private final class LazyServerStreamIterator implements Iterator<ReadObjectRespo
 
     @Override
     public boolean hasNext() {
-      return ensureResponseIteratorOpen().hasNext();
+      try {
+        return ensureResponseIteratorOpen().hasNext();
+      } catch (RuntimeException e) {
+        if (!result.isDone()) {
+          result.setException(StorageException.coalesce(e));
```
Comment (sydney-munro) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/GapicUnbufferedReadableByteChannel.java:244]: why do we set the exception on the result only if the result is not complete? What case is this covering?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/GapicUnbufferedReadableByteChannel.java [hunk 2]
```diff
@@ -237,12 +237,26 @@ private final class LazyServerStreamIterator implements Iterator<ReadObjectRespo
...
     }
 
     @Override
     public ReadObjectResponse next() {
-      return ensureResponseIteratorOpen().next();
+      try {
+        return ensureResponseIteratorOpen().next();
+      } catch (RuntimeException e) {
+        if (!result.isDone()) {
+          result.setException(StorageException.coalesce(e));
```
Comment (sydney-munro) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/GapicUnbufferedReadableByteChannel.java:256]: same question as above.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/GapicUnbufferedReadableByteChannel.java (hunk 1 above)
Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/GapicUnbufferedReadableByteChannel.java:244]: A call to `hasNext()`/`next()` could error any time. More often than not, as long as the object exists, this code will not run, but when the object doesn't exist this ensures we communicate that to the result as well as any call to `read`

File: google-cloud-storage/src/main/java/com/google/cloud/storage/GapicUnbufferedReadableByteChannel.java (hunk 2 above)
Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/GapicUnbufferedReadableByteChannel.java:256]: same answer as above

**Review by sydney-munro — APPROVED** [review]

**Pull request merged**